Перевод React.Component (#131)
reactjs/ru.reactjs.org: content/docs/reference-react-component.md

  - "tips/use-react-with-other-libraries.html"
---

Эта страница содержит подробный справочник API для определения классового компонента React. Предполагается, что вы знакомы с такими концепциями React, как [компоненты и пропсы](/docs/components-and-props.html), а также [состояние и жизненный цикл](/docs/state-and-lifecycle.html). Прочитайте про них, если вы этого не сделали.

## Обзор {#overview}

> React позволяет определять компоненты как классы или функции. В настоящее время классовые компоненты имеют больше возможностей. Они разобраны на этой странице. Чтобы определить такой компонент, необходимо расширить `React.Component`:

Author: Может отнаследоваться?

> React позволяет определять компоненты как классы или функции. В настоящее время классовые компоненты имеют больше возможностей. Они разобраны на этой странице. Чтобы определить такой компонент, необходимо расширить `React.Component`:

Reviewer: Обновил.

> React позволяет определять компоненты как классы или функции. В настоящее время классовые компоненты имеют больше возможностей. Они разобраны на этой странице. Чтобы определить такой компонент, необходимо отнаследоваться от `React.Component`:

Author: @gaearon, может убрать уточнение, что классы имеют больше возможностей? После появления хуков, мне кажется, это уже можно не уточнять. Могу создать реквест в основной документации.

Author: Пока рано (https://reactjs.org/docs/hooks-faq.html#do-hooks-cover-all-use-cases-for-classes), закрываю вопрос.

```js
class Welcome extends React.Component {
  render() {
    return <h1>Привет, {this.props.name}</h1>;
  }
}
```

Единственный *обязательный* метод в подклассе `React.Component` – [`render()`](#render). Все остальные методы, описанные ниже, являются необязательными.

**Мы рекомендуем не создавать собственные классы базовых компонентов.** В компонентах React переиспользование кода обычно достигается за счет [композиции, а не наследования](/docs/composition-vs-inheritance.html).

> Примечание:
>
>React не заставляет вас использовать синтаксис классов из ES6. Вместо этого вы можете использовать модуль `create-react-class` или его аналоги. Посмотрите раздел [Использование React без ES6](/docs/react-without-es6.html), чтобы узнать больше.

### Жизненный цикл компонента {#the-component-lifecycle}

> Каждый компонент имеет несколько «методов жизненного цикла». Переопределение этих методов позволяет выполнять код в конкретное время в процессе. **Вы можете использовать [эту диаграмму жизненного цикла](http://projects.wojtekmaj.pl/react-lifecycle-methods-diagram/) как шпаргалку.** Далее на странице **полужирным шрифтом** выделены самые распространённые методы жизненного цикла.

Reviewer: > The rest of them exist for relatively rare use cases.

```suggestion
Каждый компонент имеет несколько «методов жизненного цикла». Переопределение этих методов позволяет выполнять код в конкретное время в процессе. **Вы можете использовать [эту диаграмму жизненного цикла](http://projects.wojtekmaj.pl/react-lifecycle-methods-diagram/) как шпаргалку.** Далее на странице **полужирным шрифтом** выделены самые распространённые методы жизненного цикла. Остальные из них существуют для относительно редких вариантов использования.
```

Author: Может «остальные из них используются реже»?

Reviewer: Обновил, может так?

Author: «_тогда_ как остальные» вроде правильнее.

Изначально я не перевёл эту строчку потому что она показалась мне водой. Если распространённые выделены, то логично, что остальные нераспространённые. Может не добавлять про остальные всё таки?

Reviewer: Хорошо

- [**`render()`**](#render)
- [**`componentDidMount()`**](#componentdidmount)

> Примечание:
>
> Этот метод устарел. [Не используйте его](/blog/2018/03/27/update-on-async-rendering.html) в новом коде.
>
>- [`UNSAFE_componentWillMount()`](#unsafe_componentwillmount)

#### Обновление {#updating}

Обновление происходит при изменении пропсов или состояния. Эти методы вызываются в следующем порядке при повторном рендере компонента:

- [`static getDerivedStateFromProps()`](#static-getderivedstatefromprops)
- [`shouldComponentUpdate()`](#shouldcomponentupdate)
- [**`render()`**](#render)
- [`getSnapshotBeforeUpdate()`](#getsnapshotbeforeupdate)
- [**`componentDidUpdate()`**](#componentdidupdate)

> Примечание:
>
> Эти методы устарели. [Не используйте их](/blog/2018/03/27/update-on-async-rendering.html) в новом коде.
>
>- [`UNSAFE_componentWillUpdate()`](#unsafe_componentwillupdate)
>- [`UNSAFE_componentWillReceiveProps()`](#unsafe_componentwillreceiveprops)

#### Размонтирование {#unmounting}

Этот метод вызывается при удалении компонента из DOM:

- [**`componentWillUnmount()`**](#componentwillunmount)

#### Обработка ошибок {#error-handling}

> Эти методы вызываются если произошла ошибка в процессе рендеринга, методе жизненного цикла или конструкторе любого дочернего компонента.

Reviewer: ```suggestion
Эти методы вызываются, если произошла ошибка в процессе рендеринга, методе жизненного цикла или конструкторе любого дочернего компонента.
```

- [`static getDerivedStateFromError()`](#static-getderivedstatefromerror)
- [`componentDidCatch()`](#componentdidcatch)

> ### Другие API {#other-apis}

Reviewer: Что обозначить множественное число:

```suggestion
### Другие методы API {#other-apis}
```

Методы в этом разделе охватывают большинство задач, с которыми вы столкнётесь при использовании React-компонентов. **Для визуального представления вы можете использовать [эту диаграмму жизненного цикла](http://projects.wojtekmaj.pl/react-lifecycle-methods-diagram/).**

### `render()` {#render}

```javascript
render()
```

`render()` – единственный обязательный метод в классовом компоненте.

При вызове он проверяет `this.props` и `this.state` и возвращает один из следующих вариантов:

- **Элемент React.** Обычно создаётся с помощью [JSX](/docs/introducing-jsx.html). Указывает React, что рендерить: DOM-узел или пользовательский компонент. Например, `<div />` или `<MyComponent />`.
- **Массивы и фрагменты.** Возвращает несколько элементов из `render()`. Подробнее про [фрагменты](/docs/fragments.html).
- **Порталы**. Рендерит несколько дочерних элементов в другое поддерево DOM. Подробнее про [порталы](/docs/portals.html).
- **Строки и числа.** Рендерит текстовые DOM-узлы.
- **Booleans or `null`**. Ничего не рендерит. (Обычно необходим для поддержки паттерна `return test && <Child />`, где `test` – логическое значение.)

> Функция `render()` должна быть *чистой*. Это означает, что она: не изменяет состояние компонента, всегда возвращает один и тот же результат, не взаимодействует напрямую с браузером.

Reviewer: Странно выглядит двоеточие, так и хочется убрать

```javascript
constructor(props)
```
**Вы можете не использовать конструктор в React-компоненте, если вы не определяете состояние или не привязываете методы.**

> Конструктор компонента React вызывается до того, как компонент будет примонтирован. Вначале конструктора необходимо вызывать `super(props)`. Если это не сделать, `this.props` не будет определен. Это может привести к багам.

Reviewer: ```suggestion
Конструктор компонента React вызывается до того, как компонент будет примонтирован. В начале конструктора необходимо вызывать `super(props)`. Если это не сделать, `this.props` не будет определен. Это может привести к багам.
```

Конструкторы в React обычно используют для двух целей:

* Инициализация [внутреннего состояния](/docs/state-and-lifecycle.html) через присвоение объекта `this.state`.
* Привязка [обработчиков событий](/docs/handling-events.html) к экземпляру.

Вы **не должны вызывать `setState()`** в `constructor()`. Если вам нужно внутреннее состояние, **присвойте начальное состояние `this.state`** прямо в конструкторе.

```js
constructor(props) {
  super(props);
  // Не вызывайте здесь this.setState()!
  this.state = { counter: 0 };
  this.handleClick = this.handleClick.bind(this);
}
```

Конструктор — единственное место, где можно напрямую изменять `this.state`. В остальных методах необходимо использовать `this.setState()`.

Не используйте побочные эффекты или подписки в конструкторе. Вместо этого используйте `componentDidMount()`.

> Примечание:
>
>**Не копируйте пропсы в состояние! Это распространённая ошибка:**
>
>```js
>constructor(props) {
>  super(props);
>  // Не делайте этого!
>  this.state = { color: props.color };
>}
>```
>

> >Проблема в том, что это излишне и приводит к багам (обновления в пропе `color` не будут закоммичены в состоянии). Вместо этого используйте `this.props.color`.

Reviewer: Может быть как-то так?

```suggestion
>Проблема в том, что это излишне и приводит к багам (обновления в пропе `color` не изменят соответствующим образом состояние). Вместо этого используйте `this.props.color`.
```

Author: Может так?
обновления «не изменят состояние соответствующим образом»
обновления «не будут зафиксированы в состоянии»

Reviewer: Второй вариант, мне кажется лучше

componentDidMount()
```

`componentDidMount()` вызывается сразу после монтирования (компонент вставлен в DOM). В этом методе должны происходить действия, которые требуют наличия DOM-узлов. Это хорошее место для создания сетевых запросов.

Этот метод подходит для настройки подписок. Но не забудьте отписаться от них в `componentWillUnmount()`.

> Вы **можете сразу вызвать setState()** в `componentDidMount()`. Это вызовет дополнительный рендер перед тем как браузер обновит экран. Гарантируется, что пользователь не увидит промежуточное состояние, даже если `render()` будет вызываться дважды. Используйте этот подход с осторожностью, он может вызвать проблемы с производительностью. В большинстве случаев начальное состояние лучше объявить в `constructor()`. Однако, это может быть необходимо для случаев, когда нужно измерить размер или положение DOM-узла на основе которого происходит рендер. Например, для модальных окон или всплывающих подсказок.

Reviewer 2: На мой взгляд, здесь можно добавить запятых.

```suggestion
Вы **можете сразу вызвать setState()** в `componentDidMount()`. Это вызовет дополнительный рендер перед тем, как браузер обновит экран. Гарантируется, что пользователь не увидит промежуточное состояние, даже если `render()` будет вызываться дважды. Используйте этот подход с осторожностью, он может вызвать проблемы с производительностью. В большинстве случаев начальное состояние лучше объявить в `constructor()`. Однако, это может быть необходимо для случаев, когда нужно измерить размер или положение DOM-узла, на основе которого происходит рендер. Например, для модальных окон или всплывающих подсказок.
```

componentWillUnmount()
```

`componentWillUnmount()` вызывается сразу перед размонтированием и удалением компонента. В этом методе выполняется необходимый сброс: отмена таймеров, сетевых запросов и подписок, созданных в `componentDidMount()`.

**Не используйте setState()** в `componentWillUnmount()`, так как компонент никогда не рендерится повторно. После того, как экземпляр компонента будет размонтирован, он никогда не будет примонтирован снова.

* * *

### Редко используемые методы жизненного цикла {#rarely-used-lifecycle-methods}

> Методы из этого раздела используются редко. В большинстве компонентов они не нужны, хотя иногда бывают полезны. **Вы можете увидеть большинство приведённых ниже методов на [этой диаграмме жизненного цикла](http://projects.wojtekmaj.pl/react-lifecycle-methods-diagram/), если наверху страницы нажмете на чекбокс «Show less common lifecycles».**

Reviewer: > Show less common lifecycles

TODO: поменять на "Показать менее популярные методы жизненного цикла", когда будет принят PR https://github.com/wojtekmaj/react-lifecycle-methods-diagram/pull/36

Reviewer: ```suggestion
Методы из этого раздела используются редко. В большинстве компонентов они не нужны, хотя иногда бывают полезны. **Вы можете увидеть большинство приведённых ниже методов на [этой диаграмме жизненного цикла](http://projects.wojtekmaj.pl/react-lifecycle-methods-diagram/), если наверху страницы нажмете на чекбокс «Показать менее популярные методы жизненного цикла», выбрав предварительно русский язык.**
```

### `shouldComponentUpdate()` {#shouldcomponentupdate}

```javascript
shouldComponentUpdate(nextProps, nextState)
```

Используйте `shouldComponentUpdate()`, чтобы указать необходимость следующего рендера на основе изменений состояния и пропсов. По умолчанию происходит повторный рендер при любом изменении состояния. В большинстве случаев вы должны полагаться на это поведение.

`shouldComponentUpdate()` вызывается перед рендером, когда получает новые пропсы или состояние. Значение по умолчанию равно `true`. Этот метод не вызывается при первом рендере или когда используется `forceUpdate()`.

Этот метод нужен только для **[повышения производительности](/docs/optimizing-performance.html).**. Но не опирайтесь на его возможность «предотвратить» рендер, это может привести к багам. Вместо этого **используйте [`PureComponent`](/docs/react-api.html#reactpurecomponent)**, который позволяет не описывать поведение `shouldComponentUpdate()` вручную. `PureComponent` поверхностно сравнивает пропсы и состояние и позволяет не пропустить необходимое обновление.

> Если вы уверены, что хотите написать его вручную, вы можете сравнить `this.props` с `nextProps` и `this.state` с `nextState`. Верните `false` что бы пропустить обновление React. Возвращение `false` не предотвращает повторный рендер дочерних компонентов при изменении *их* состояния.

Reviewer: ```suggestion
Если вы уверены, что хотите написать его вручную, вы можете сравнить `this.props` с `nextProps`, а `this.state` с `nextState`. Верните `false` чтобы пропустить обновление React. Возврат `false` не предотвращает повторный рендер дочерних компонентов при изменении *их* состояния.
```

static getDerivedStateFromProps(props, state)
```

> `getDerivedStateFromProps` вызывается сразу перед вызовом метода `render`, как при начальном монтировании, так и при последующих обновлениях. Он должен вернуть объект для обновления состояния или null, чтобы ничего не обновлять.

Reviewer 3: ```suggestion
`getDerivedStateFromProps` вызывается непосредственно перед вызовом метода `render`, как при начальном монтировании, так и при последующих обновлениях. Он должен вернуть объект для обновления состояния или null, чтобы ничего не обновлять.
```

> Этот метод существует для [редких случаев](/blog/2018/06/07/you-probably-dont-need-derived-state.html#when-to-use-derived-state), когда изменения в пропсах не сразу влияют на состояние. Например, это подойдёт для реализации компонента `<Transition>`, который сравнивает свои предыдущие и следующие дочерние компоненты, чтобы решить, какой из них нужно анимировать.

Reviewer: А почему не _когда состояние зависит от изменений в пропсах..._?

> когда изменения в пропсах не сразу влияют на состояние
> where the state depends on changes in props over time

Author: Наверно меня смутил over time. Исправлю на «когда состояние зависит от изменений в пропсах»

Производное состояние приводит к сложному коду и делает ваши компоненты сложными для понимания. Убедитесь, что вы знакомы с [простыми альтернативами:](/blog/2018/06/07/you-probably-dont-need-derived-state.html)

* Чтобы **выполнить побочный эффект при изменении пропсов** (например, сетевой запрос или анимацию) используйте [`componentDidUpdate`](#componentdidupdate).

* Чтобы **повторно вычислить данные при изменении пропсов**, используйте [функцию мемоизации](/blog/2018/06/07/you-probably-dont-need-derived-state.html#what-about-memoization).

* Чтобы **«сбросить» некоторое состояние при изменении пропсов**, используйте [контролируемые компоненты](/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-controlled-component) или [неконтролируемые компоненты с ключом](/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-uncontrolled-component-with-a-key).

Этот метод не имеет доступа к экземпляру компонента. Чтобы использовать пропсы, состояние и методы класса в `getDerivedStateFromProps()`, их нужно вынести за пределы класса в виде чистых функций.

> Если вы хотите переиспользовать код между `getDerivedStateFromProps()` и другими методами класса, извлекая чистые функции пропсов и состояния компонента вне определения класса.

Reviewer 3: ```suggestion
Если вы хотите переиспользовать код между `getDerivedStateFromProps()` и другими методами класса, извлеките чистые функции пропсов и состояния компонента и поместите их вне определения класса.
```

getSnapshotBeforeUpdate(prevProps, prevState)
```

`getSnapshotBeforeUpdate()` вызывается прямо перед этапом «фиксирования» (например, перед добавлением в DOM). Он позволяет вашему компоненту брать некоторую информацию из DOM (например, положение прокрутки) перед её возможным изменением. Любое значение, возвращаемое этим методом жизненного цикла, будет передано как параметр `componentDidUpdate()`.

> Это применяется редно, но может быть полезно в таких интерфейсах, как цепочка сообщений в чатах, в которых позиция прокрутки обрабатывается особым образом.

Reviewer 2: ```suggestion
Это применяется редко, но может быть полезно в таких интерфейсах, как цепочка сообщений в чатах, в которых позиция прокрутки обрабатывается особым образом.
```

Значение снимка (или `null`) должно быть возвращено.

Например:

`embed:react-component-reference/get-snapshot-before-update.js`

В примерах выше важно получить значение свойства `scrollHeight` в `getSnapshotBeforeUpdate` из-за того, что могут возникать задержки между этапами жизненного цикла «рендер» (например, `render`) и «фиксирование» (например, `getSnapshotBeforeUpdate` и `componentDidUpdate`).

* * *

### Предохранители {#error-boundaries}

> [Предохранители](/docs/error-boundaries.html) – это React-компоненты, которые перехватывают JavaScript-ошибки в любом месте их дочернего дерева компонентов. Затем логируют эти ошибки и отображают запасной интерфейс вместо разрушенного дерева компонентов. Предохранители отлавливают ошибки при рендере, в методах жизненного цикла и в конструкторах всего дерева под ними.

Reviewer 3: ```suggestion
[Предохранители](/docs/error-boundaries.html) – это React-компоненты, которые перехватывают JavaScript-ошибки в любом месте их дочернего дерева компонентов. Затем логируют эти ошибки и отображают запасной интерфейс вместо "поломанного" дерева компонентов. Предохранители отлавливают ошибки при рендере, в методах жизненного цикла и в конструкторах всего дерева под ними.
```

Классовый компонент становится предохранителем, если в нём используются методы жизненного цикла `static getDerivedStateFromError()` и (или) `componentDidCatch()`. Обновление состояния в этом методе жизненного цикла позволяет перехватить необработанную JavaScript-ошибку в дереве ниже и отобразить запасной интерфейс.

> Используйте предохранители только для обработки неожиданных исключений, **не используйте их для управления потоком.**

Reviewer 3: ```suggestion
Используйте предохранители только для обработки неожиданных исключений, **не используйте их для управления потоком исполнения в вашем приложении.**
```

  render() {
    if (this.state.hasError) {
      // Здесь можно рендерить запасной интерфейс
      return <h1>Something went wrong.</h1>;
    }

    return this.props.children;
  }
}
```

> Примечание:
>
>В случае ошибки вы можете рендерить запасной интерфейс с помощью `componentDidCatch()`, вызвав `setState`. Однако, этот способ скоро будет считаться устаревшим.
> Используйте `static getDerivedStateFromError()` для рендера резервного интерфейса.

* * *

### Устаревшие методы жизненного цикла {#legacy-lifecycle-methods}

Приведённые ниже методы жизненного цикла устарели. Их не рекомендуется использовать в новом коде, хотя они продолжают работать. Вы можете узнать больше о переходе с устаревших методов жизненного цикла в [блоге](/blog/2018/03/27/update-on-async-rendering.html).

### `UNSAFE_componentWillMount()` {#unsafe_componentwillmount}

```javascript
UNSAFE_componentWillMount()
```

> Примечание:
>
> Этот метод жизненного цикла раньше назывался `componentWillMount`. По этому названию он будет доступен до 17 версии. Чтобы автоматически обновить компоненты, используйте [`rename-unsafe-lifecycles`](https://github.com/reactjs/react-codemod#rename-unsafe-lifecycles).

> `UNSAFE_componentWillMount()` вызывается сразу перед монтированием. Он вызывается перед `render()`, поэтому синхронный вызов `setState()` в этом методе не вызовет дополнительный рендер. Для инициализации состояния мы рекомендуем использовать `constructor()`.

Reviewer 3: ```suggestion
`UNSAFE_componentWillMount()` вызывается непосредственно перед монтированием. Он вызывается перед `render()`, поэтому синхронный вызов `setState()` в этом методе не вызовет дополнительный рендер. Для инициализации состояния мы рекомендуем использовать `constructor()`.
```

UNSAFE_componentWillReceiveProps(nextProps)
```

> Примечание:
>
> Этот метод жизненного цикла раньше назывался `componentWillReceiveProps`. По этому названию он будет доступен до 17 версии. Чтобы автоматически обновить компоненты, используйте [`rename-unsafe-lifecycles`](https://github.com/reactjs/react-codemod#rename-unsafe-lifecycles).

> Примечание:
>

> > Частое использование этого метода жизненного цикла приводило к багам

Reviewer 3: ```suggestion
> Использование этого метода жизненного цикла часто приводило к багам
```

UNSAFE_componentWillUpdate(nextProps, nextState)
```

> Примечание:
>
> Этот метод жизненного цикла раньше назывался `componentWillUpdate`. По этому названию он будет доступен до 17 версии. Чтобы автоматически обновить компоненты, используйте [`rename-unsafe-lifecycles`](https://github.com/reactjs/react-codemod#rename-unsafe-lifecycles).

> `UNSAFE_componentWillUpdate()` вызывается сразу перед рендером при получении новых пропсов или состояния. В этом методе можно выполнить некоторую подготовку перед обновлением. Этот метод не вызывается при первом рендере.

Reviewer 3: ```suggestion
`UNSAFE_componentWillUpdate()` вызывается непосредственно перед рендером при получении новых пропсов или состояния. В этом методе можно выполнить некоторую подготовку перед обновлением. Этот метод не вызывается при первом рендере.
```

> Внутри этого метода нельзя вызвать `this.setState()`, а также делать какие-либо действия, которые влияют на обновление компонента перед возвратом `UNSAFE_componentWillUpdate()` (например, вызывать `Redux action`).

Reviewer: ```suggestion
Внутри этого метода нельзя вызвать `this.setState()`, а также делать какие-либо действия, которые влияют на обновление компонента перед возвратом `UNSAFE_componentWillUpdate()` (например, отправка действия Redux).
```

> Обычно, этот метод можно заменить на `componentDidUpdate()`. Вы можете использовать `getSnapshotBeforeUpdate()` для работы с DOM (например, запоминать положение прокрутки страницы).

Reviewer: ```suggestion
Обычно этот метод можно заменить на `componentDidUpdate()`. Вы можете использовать `getSnapshotBeforeUpdate()` для работы с DOM (например, запоминать положение прокрутки страницы).
```

> Примечание:
>
> `UNSAFE_componentWillUpdate()` не вызывается, если [`shouldComponentUpdate()`](#shouldcomponentupdate) возвращает `false`.

* * *

## Другие APIs {#other-apis-1}

В отличие от методов жизненного цикла, представленных выше (React вызывает их сам), методы, приведённые ниже, можно вызывать из компонентов.

Их всего два: `setState()` и `forceUpdate()`.

### `setState()` {#setstate}

```javascript
setState(updater[, callback])
```

`setState()` добавляет в очередь изменения в состоянии компонента. Также он указывает React, что компонент и его дочерние элементы должны быть повторно отрендерены с обновлённым состоянием. Этот метод используется для обновления интерфейса в ответ на обработчики событий и ответы сервера.

Думайте о `setState()`, как о *запросе*, а не как о команде немедленного обновления компонента. Для увеличения производительности React может задержать его выполнение, а затем обновить несколько компонентов за один проход. React не гарантирует моментальное применение изменений в состоянии.

> Метод `setState()` не всегда обновляет компонент сразу. Он может группировать или откладывать обновление до следующего раза. Это делает чтение `this.state` сразу после вызова `setState()` потенциальной ловушки. Вместо этого используйте `componentDidUpdate()` или колбэк `setState()` (`setState(updater, callback)`). Любой из этих вариантов можно запустить для применения обновления. Если вам нужно обновить состояние на основе предыдущего, используйте аргумент `updater`, описанный ниже.

Reviewer 3: ```suggestion
Метод `setState()` не всегда обновляет компонент сразу. Он может группировать или откладывать обновление до следующего раза. Это делает чтение `this.state` сразу после вызова `setState()` потенциальной ловушкой. Вместо этого используйте `componentDidUpdate()` или колбэк `setState()` (`setState(updater, callback)`), каждый из которых гарантированно вызывается после того как было применено обновление. Если вам нужно обновить состояние на основе предыдущего, используйте аргумент `updater`, описанный ниже.
```

> `setState()` не всегда обновляет компонент сразу. Этот метод может группировать или откладывать обновление до следующего раза. Это делает чтение `this.state` сразу после вызова `setState()` потенциальной ловушки. Вместо этого используйте `componentDidUpdate()` или колбек `setState()` (`setState(updater, callback)`). Любой из этих вариантов можно запустить для применения обновления. Если вам нужно обновить состояние на основе предыдущего, используйте аргумент `updater`, описанный ниже.

Reviewer: ```suggestion
Метод `setState()` не всегда обновляет компонент сразу. Он может группировать или откладывать обновление до следующего раза. Это делает чтение `this.state` сразу после вызова `setState()` потенциальной ловушки. Вместо этого используйте `componentDidUpdate()` или колбэк `setState()` (`setState(updater, callback)`). Любой из этих вариантов можно запустить для применения обновления. Если вам нужно обновить состояние на основе предыдущего, используйте аргумент `updater`, описанный ниже.
```

> `setState()` всегда приводит к повторному рендеру, пока `shouldComponentUpdate()` не возвращает `false`. Если используются мутабельные объекты, и условие рендеринга не может быть реализовано в `shouldComponentUpdate()`, вызывайте `setState()` только при разнице следующего и предыдущего состояния. Это предотвратит ненужные повторные рендеры.

Reviewer 3: ```suggestion
`setState()` всегда приводит к повторному рендеру, если только `shouldComponentUpdate()` не возвращает `false`. Если используются мутабельные объекты, и условие рендеринга не может быть реализовано в `shouldComponentUpdate()`, вызывайте `setState()` только при разнице следующего и предыдущего состояния. Это предотвратит ненужные повторные рендеры.
```

Первым аргументом передаётся функция `updater`, которая имеет следующий вид:

```javascript
(state, props) => stateChange
```

> `state` – ссылка на состояние компонента при изменении. Она не должна мутировать. Изменения должны быть в виде нового объекта на основе входных данных из `state` и` props`. Предположим, что мы хотели бы увеличить значение состояния с помощью `props.step`:

Reviewer 3: ```suggestion
`state` – ссылка на состояние компонента при изменении. Объект состояния не должен мутировать. Изменения должны проявляться в виде нового объекта на основе входных данных из `state` и` props`. Предположим, что мы хотели бы увеличить значение состояния с помощью `props.step`:
```

```javascript
this.setState((state, props) => {
  return {counter: state.counter + props.step};
});
```

> И `state`, и` props`, полученные функцией обновления, гарантированно будут обновлены. Результат функции поверхностно объединяется с `state`.

Reviewer: ```suggestion
Как `state`, так и `props`, полученные функцией обновления, гарантированно будут обновлены. Результат функции поверхностно объединяется с `state`.
```

> Второй параметр в `setState()` - дополнительная колбек-функция, которая выполняется после того, как исполнится `setState` и произойдет повторный рендер компонента. Мы рекомендуем использовать такой подход в `componentDidUpdate()`.

Reviewer: Упростить можно так:

```suggestion
Второй параметр в `setState()` - дополнительный колбэк, который выполняется после того, как исполнится `setState` и произойдет повторный рендер компонента. Мы рекомендуем использовать такой подход в `componentDidUpdate()`.
```

> В качестве первого аргументы `setState()`, вместо функции, вы можете передать объект:

Reviewer: ```suggestion
В качестве первого аргумента `setState()`, вместо функции, вы можете передать объект:
```

component.forceUpdate(callback)
```

> По умолчанию, при изменении состояния компонента или пропсов, происходит повторный рендер. Если ваш метод `render()` зависит от некоторых других данных, вы можете указать React необходимость в повторном рендере, вызвав `forceUpdate()`.

Reviewer: ```suggestion
По умолчанию при изменении состояния компонента или пропсов, происходит повторный рендер. Если ваш метод `render()` зависит от некоторых других данных, вы можете указать React необходимость в повторном рендере, вызвав `forceUpdate()`.
```

Вызов `forceUpdate()` приведёт к выполнению метода `render()` в компоненте, пропуская `shouldComponentUpdate()`. Это вызовет обычные методы жизненного цикла для дочерних компонентов, включая `shouldComponentUpdate()` каждого дочернего компонента. React по-прежнему будет обновлять DOM только в случае изменения разметки.

Чаще всего, `forceUpdate()` не используется. Вместо этого используются в `render()` данные из `this.props` и `this.state`.

* * *

## Свойства класса {#class-properties-1}

### `defaultProps` {#defaultprops}

> `defaultProps` можно определить как свойство самого класса компонента, которое позволяет установить пропсы класса по умолчанию. Это используется для неопределённых пропсов, но не для пропсов со значением null. Например:

Reviewer 3: ```suggestion
`defaultProps` можно определить как свойство самого класса компонента, которое позволяет установить пропсы класса по умолчанию. Это используется для неопределённых (`undefined`) пропсов, но не для пропсов со значением `null`. Например:
```

### `displayName` {#displayname}

Строка `displayName` используется в сообщениях для отладки. Обычно вам не нужно её явно указывать. По умолчанию используется имя функции или класса, указанное при определении компонента. Если вам нужно установить его явно, например, для отладки или создания компонента высшего порядка, посмотрите раздел [Обертка отображаемого имени для простой отладки](/docs/higher-order-components.html#convention-wrap-the-display-name-for-easy-debugging).

* * *

## Свойства экземпляра {#instance-properties-1}

### `props` {#props}

> `this.props` содержит свойства, которые были определены вызывающим элементом этого компонента. Подробнее об этом можно узнать в разделе [Компоненты и пропсы](/docs/components-and-props.html)

Reviewer 3: ```suggestion
`this.props` содержит свойства, которые были определены тем, кто вызывает этот компонента. Подробнее об этом можно узнать в разделе [Компоненты и пропсы](/docs/components-and-props.html)
```

Существует специальный проп `this.props.children`, который обычно определяется дочерними тегами в JSX-выражении, а не в самом теге.

### `state` {#state}

> Состояние содержит данные, конкретные для этого компонента. Они могут измениться со временем. Состояние определяется пользователем. Оно должно быть простым объектом JavaScript.

Reviewer 3: ```suggestion
Состояние содержит данные, специфичные для этого компонента. Они могут измениться со временем. Состояние определяется пользователем и должно быть простым объектом JavaScript.
```

> Вам не нужно вставлять в состояние значение, которое не используется для рендера или потока данных (например, идентификатор таймера). Такие значения можно определить как поля экземпляра компонента.

Reviewer 3: ```suggestion
Вам не нужно вставлять в состояние значение, если оно не используется для рендера или потока данных (например, идентификатор таймера). Такие значения можно определить как поля экземпляра компонента.
```

Дополнительную информацию можно найти в разделе [Состояние и жизненный цикл](/docs/state-and-lifecycle.html).

> Никогда не мутируйте `this.state` напрямую, так как позднее вызов `setState()` может заменить мутацию, которую вы сделали. Относитесь к `this.state` как к иммутабельному объекту.

Reviewer 3: ```suggestion
Никогда не мутируйте `this.state` напрямую, так как более поздний вызов `setState()` может заместить эту мутацию. Относитесь к `this.state` как к иммутабельному объекту.
```

Author: Может лучше «перезаписать»?
